### PR TITLE
Created Hashtable, validation, and integration with other teams

### PR DESCRIPTION
**Pull request by ecoble**
This PR includes:

- Creation of a hashtable (using UTHASH) that maps commands to a corresponding operation and action struct. Fills the table with action commands from a list specified by action management, as well "other" commands such as "help", "quit", "save", etc.

- Preliminary validation of commands. We verify that actions exist through the hashtable, if the action is not found in the hashtable an error is produced. Objects are verified through a game-state function that checks if an object is in the room. We do not check if an action can be performed on an object once the action and object have both been verified, this is action management's responsibility.

- Integration of our code with game state and action management. We now use game state's structs and macros to check whether objects are in a room and to get descriptions of rooms and objects when the command "LOOK" is called, and to get inventory when "INV" is called. We also pass on actions and objects to action management's functions  do_item_action, do_path_action, and do_item_item_action depending on the input the user gives.

- Parser input sanitization. Parser now gives an error and returns NULL if the user types too many words, and it also capitalizes the entire first word (the command word) before outputting the token list.

- Began integration with checkpointing. Starting integrating the save into our code, but currently there is no working save function

Considerations:

- We have not yet integrated the hashtable with the context structs, this will be the next task we work on. This will cause a lot of changes to parameter types, mostly everywhere we pass game_t will have to be changed to a context struct

- Currently in our shell calling "LOOK" or any other game state related function will cause a Segfault, this is expected since there is no room, but maybe we should add a catch so this doesn't happen

- The function that populates the hashtable with action management's actions does not work currently, but it will work once action management changes the capitalization in their list of valid actions. (We have requested commands be all uppercase since this is the standard we have been following, and we cannot currently accept two word commands, e.g. turn on)

### CODE REVIEW

**Comment by ecoble** [comment]
@Gestalt-R Where do you want more comments? Sorry I may have misunderstood you when we talked earlier, I thought we discussed updating comments to have the correct parameters for functions, which I did do

**Comment by Gestalt-R** [comment]
I just meant that I normally feel obligated to add feedback, but here, I have nothing to add! It's good to go!

File: src/action_management/Makefile [hunk 1]
```diff
@@ -24,4 +24,4 @@ $(SRCS:.c=.d):%.d:%.c
 .PHONY: clean
 
 clean:
-	${RM} ${LIB} ${OBJS} $(SRCS:.c=.d)
\ No newline at end of file
+	${RM} ${LIB} ${OBJS} $(SRCS:.c=.d)
```
Comment (lydiafilipe) [inline src/action_management/Makefile]: It looks like you are adding tabs here, please make sure to only be using spaces

File: src/cli/Makefile [hunk 2]
```diff
@@ -11,16 +11,13 @@ RM = rm -f
 all: $(LIB)
 
 $(LIB): $(OBJS)
-	ar -r -o $@ $(OBJS) 
+	ar -r -o $@ $(OBJS)
```
Comment (lydiafilipe) [inline src/cli/Makefile]: same here about tabs

File: src/cli/include/cmd.h [hunk 3]
```diff
@@ -1,21 +1,125 @@
 #ifndef _CLI_INCLUDE_CMD_H
```
Comment (lydiafilipe) [inline src/cli/include/cmd.h]: some of the spacing in this file is inconsistent. Please leave two blank lines after function declarations and none after function comments

File: src/cli/src/shell.c [hunk 4]
```diff
@@ -13,30 +13,30 @@ void help_text()
     char p[] =
```
Comment (lydiafilipe) [inline src/cli/src/shell.c]: there are mixes of tabs and spaces in here, please only be using spaces

File: src/cli/src/operations.c [hunk 5]
```diff
@@ -4,72 +4,147 @@
...
-    printf("%s\n",tokens[0] );
-    return "is a direction action!";
+    if(tokens[1]==NULL)
+    {
+        return game->curr_room->long_desc;
+    }
+    item_t *curr_item;
+    ITER_ALL_ITEMS_IN_ROOM(game->curr_room, curr_item)
+    {
+        if (strcmp(curr_item->item_id,tokens[1])==0)
```
Comment (lydiafilipe) [inline src/cli/src/operations.c]: small thing but make sure there's spaces around the ==, there are quite a few places where that isn't the case in this file

File: src/cli/src/parser.c [hunk 6]
```diff
@@ -23,5 +24,38 @@ char **parse(char *input)
...
+    {
+        ch = toupper(command[i]);
+        command[i] = ch;
+        i++;
+    }
+
     return words;
 }
+
+//A test function
```
Comment (lydiafilipe) [inline src/cli/src/parser.c]: please remove commented out code

File: src/cli/include/cmd.h [hunk 7]
```diff
@@ -1,21 +1,125 @@
...
+ *
+ * Parameters:
+ * - command name, pointer to table
+ *
+ * Returns:
+ * - a pointer to the corresponding action
+ * - NULL if the command is not an action.
+ */
+action_type_t *find_action(char *command_name, lookup_t **table);
+/*  Commented out until action_t is ready.*/
```
Comment (lydiafilipe) [inline src/cli/include/cmd.h]: Is this comment supposed to be here?

File: src/cli/src/cmd.c [hunk 8]
```diff
@@ -5,14 +5,101 @@
...
+{
+    lookup_t *t;
+    HASH_FIND_STR(*table, command_name, t);
+    return t;
+}
+
+operation *find_operation(char *command_name, lookup_t **table)
+{
+    lookup_t *t;
+    if ((t = find_entry(command_name, table))) return t->operation_type;
```
Comment (lydiafilipe) [inline src/cli/src/cmd.c]: Please have the return statement indented on a new line, and use braces in the if statement

File: src/cli/src/cmd.c [hunk 9]
```diff
@@ -24,10 +111,10 @@ cmd *cmd_new(char *tokens[TOKEN_LIST_SIZE])
 /* See cmd.h */
 void cmd_free(cmd *c)
 {
-    if (c == NULL || c->tokens == NULL) return;
+    if(c == NULL || c->tokens == NULL) return;
```
Comment (lydiafilipe) [inline src/cli/src/cmd.c]: Please have the return statement indented on a new line, and use braces in the if statement

File: src/cli/src/cmd.c [hunk 10]
```diff
@@ -24,10 +111,10 @@ cmd *cmd_new(char *tokens[TOKEN_LIST_SIZE])
...
     for(int i = 0; i < TOKEN_LIST_SIZE; i++)
     {
-        if (c->tokens[i] != NULL) free(c->tokens[i]);
+        if(c->tokens[i] != NULL) free(c->tokens[i]);
```
Comment (lydiafilipe) [inline src/cli/src/cmd.c]: Please have the free statement indented on a new line, and use braces in the if statement

File: src/cli/src/cmd.c [hunk 11]
```diff
@@ -5,14 +5,101 @@
...
 
 /* See cmd.h */
 cmd *cmd_new(char *tokens[TOKEN_LIST_SIZE])
 {
     cmd *c = (cmd*)malloc(sizeof(cmd));
-    if (c==NULL)
+    if(c == NULL)
     {
         fprintf(stderr,"error (cmd_tag): malloc failed\n");
         exit(1);
```
Comment (lydiafilipe) [inline src/cli/src/cmd.c]: Sorry I didn't catch before, but why exit instead of return null?

Comment (ecoble) [reply src/cli/src/cmd.c]: This may have been a remnant of code that came from a shell from 152, I'll change it to return null

**Comment by ecoble** [comment]
@lydiafilipe In regards to tests, we haven't started writing any yet. Writing tests will be new tasks for Sprint 4.

File: src/action_management/Makefile (hunk 1 above)
Comment (ecoble) [reply src/action_management/Makefile]: I tried erasing each tab and replacing it with four spaces, but after I did this the Makefile kept failing with the error "*** missing separator. Stop" I googled it and I think Makefiles requires tabs for the action of every rule.

Here are the links I looked at:
https://stackoverflow.com/questions/16931770/makefile4-missing-separator-stop
https://stackoverflow.com/questions/2131213/can-you-make-valid-makefiles-without-tab-characters

**Comment by ecoble** [comment]
To get the build to pass, I edited UI's Makefile to include action management's header files, but I was told by Borja that this was okay. There were also some functions UI starting using of ours that have since been updated, so I had to fill some parameters with NULL there. I am working on a new branch to integrate UI and CLI so hopefully that will be fixed soon.

File: src/cli/include/operations.h [hunk 12]
```diff
@@ -16,109 +19,154 @@
...
+ * - tokens: parsed input string (validified)
+ *  - pointer to game
+ *  - pointer to table
+ * 
  * Returns:
- * - error message as a string
+ * - Said list of supported operations as a string
  */
-char *prep_error_operation(char *tokens[TOKEN_LIST_SIZE]);
+char *type1_action_operation(char *tokens[TOKEN_LIST_SIZE], game_t * game, lookup_t **table);
```
Comment (SnowballAntrobus) [inline src/cli/include/operations.h]: The use of type1 is inconsistent with the convention established by action management. Actions such as "eat", "walk", etc are types of actions. We use either kind1_action or ITEM_action to denote this family of actions

File: src/cli/include/operations.h [hunk 13]
```diff
@@ -16,109 +19,154 @@
...
  * Parameters:
- * - tokens: parsed input string
+ * - tokens: parsed input string (validified)
+ *  - pointer to game
+ *  - pointer to table
+ * 
+ * Returns:
+ * - Said list of supported operations as a string
+ */
+char *type2_action_operation(char *tokens[TOKEN_LIST_SIZE], game_t * game, lookup_t **table);
```
Comment (SnowballAntrobus) [inline src/cli/include/operations.h]: The use of type2 is inconsistent with the convention established by action management. Actions such as "eat", "walk", etc are types of actions. We use either kind2_action or PATH_action to denote this family of actions

File: src/cli/include/operations.h [hunk 14]
```diff
@@ -16,109 +19,154 @@
...
+ *  - pointer to game
+ *  - pointer to table
+ * 
  * Returns:
- * - error message as a string
- */                                                       
-char *ind_object_error_operation(char *tokens[TOKEN_LIST_SIZE]);
+ * - Said list of supported operations as a string
+ */
+char *type3_action_operation(char *tokens[TOKEN_LIST_SIZE], game_t * game, lookup_t **table);
```
Comment (SnowballAntrobus) [inline src/cli/include/operations.h]: The use of type3 is inconsistent with the convention established by action management. Actions such as "eat", "walk", etc are types of actions. We use either kind3_action or ITEM_ITEM_action to denote this family of actions

File: src/cli/include/operations.h [hunk 15]
```diff
@@ -16,109 +19,154 @@
...
+
 
-// Put more operations here!
+//char *type4_action_operation(char *tokens[TOKEN_LIST_SIZE], game_t * game);
```
Comment (SnowballAntrobus) [inline src/cli/include/operations.h]: Same as the above use kind4

File: src/cli/src/operations.c [hunk 16]
```diff
@@ -4,72 +4,147 @@
...
 }
-char *type4_action_operation(char *tokens[TOKEN_LIST_SIZE])
+
+//KIND 2:   ACTION <direction>
+char *type2_action_operation(char *tokens[TOKEN_LIST_SIZE], game_t *game, lookup_t **table)
 {
-    printf("%s\n",tokens[0] );
-    return "is an action!";
+    path_t *curr_path;
+    ITER_ALL_PATHS(game->curr_room, curr_path)
```
Comment (SnowballAntrobus) [inline src/cli/src/operations.c]: So I've been confused on this for a while. The player is gonna say a direction with the go/walk command so how is it you get a path from that? As far as I can see you would have to check to see what (if anything) is in that direction and then get the path (both with GS functions). What am I missing?

Comment (ecoble) [reply src/cli/src/operations.c]: The user will give a direction, say "EAST". We use game state's functions to search the room for all possible paths out of that room, and if any of them are EAST, the user is able to move in that direction, and we pass on the direction and game to action management. If not we return an "error" message. Does that clear things up?

Comment (SnowballAntrobus) [reply src/cli/src/operations.c]: Sweet! That makes sense to me.

**Comment by ecoble** [comment]
CLOSING STATEMENT: Our code now includes a functional hashtable for matching commands to operations, preliminary validation of functions, integration with action management, parser input sanitization, and the beginning on integration with checkpointing. Changes made since the initial pull request include style fixes and naming convention fixes.

**Comment by lydiafilipe** [comment]
Issue Score: ✓+

Comments:
Great work!

**Pull request merged**